File: cliente_go/cliente.go
package main

import (
	"bufio"
	"context" 
	"crypto/rand"
	"encoding/json"
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"

	amqp "github.com/rabbitmq/amqp091-go"
)

func failOnError(err error, msg string) {
	if err != nil {
		log.Panicf("%s: %s", msg, err)
	}
}


func randomString(l int) string {
	bytes := make([]byte, l)
	rand.Read(bytes)
	return fmt.Sprintf("%x", bytes)
}

func readStr(reader *bufio.Reader, prompt string) string {
	fmt.Print(prompt)
	text, _ := reader.ReadString('\n')
	return strings.TrimSpace(text)
}

func readFloat(reader *bufio.Reader, prompt string) float64 {
	str := readStr(reader, prompt)
	val, _ := strconv.ParseFloat(str, 64)
	return val
}

func readInt(reader *bufio.Reader, prompt string) int {
	str := readStr(reader, prompt)
	val, _ := strconv.Atoi(str)
	return val
}

func main() {
	// Conecta ao RabbitMQ
	conn, err := amqp.Dial("amqp://guest:guest@localhost:5672/")
	failOnError(err, "Falha ao conectar no RabbitMQ")
	defer conn.Close()

	ch, err := conn.Channel()
	failOnError(err, "Falha ao abrir o canal")
	defer ch.Close()

	
	q, err := ch.QueueDeclare("", false, false, true, false, nil)
	failOnError(err, "Falha ao declarar a fila de callback")

	msgs, err := ch.Consume(q.Name, "", true, false, false, false, nil)
	failOnError(err, "Falha ao registrar consumidor")

	reader := bufio.NewReader(os.Stdin)

	
	enviarParaFila := func(acao string, dados map[string]interface{}) map[string]interface{} {
		corrId := randomString(32)
		payloadMap := map[string]interface{}{"acao": acao, "dados": dados}
		payloadBytes, _ := json.Marshal(payloadMap)

		
		err = ch.PublishWithContext(
			context.Background(),
			"",
			"fila_banco",
			false,
			false,
			amqp.Publishing{
				DeliveryMode: amqp.Persistent,
				ContentType:   "application/json",
				CorrelationId: corrId,
				ReplyTo:       q.Name,
				Body:          payloadBytes,

			})
		failOnError(err, "Falha ao publicar a mensagem")

		// Aguarda a resposta correta chegar
		for d := range msgs {
			if corrId == d.CorrelationId {
				var res map[string]interface{}
				json.Unmarshal(d.Body, &res)
				return res
			}
		}
		return nil
	}

	for {
		fmt.Println("\n--- BANCO RN (Cliente MQ / GO) ---")
		fmt.Println("1 - Criar conta")
		fmt.Println("2 - Depositar")
		fmt.Println("3 - Sacar")
		fmt.Println("4 - Consultar saldo")
		fmt.Println("5 - Transferir")
		fmt.Println("6 - Sair")

		opcao := readStr(reader, "\nEscolha: ")

		switch opcao {
		case "1":
			login := readStr(reader, "Login: ")
			nome := readStr(reader, "Nome: ")
			tipo := readStr(reader, "Tipo de conta (corrente/poupanca): ")
			numero := readInt(reader, "Numero da conta: ")
			saldo := readFloat(reader, "Saldo inicial: ")

			var limite, taxa float64
			if tipo == "corrente" {
				limite = readFloat(reader, "Limite cheque especial: ")
			} else {
				taxa = readFloat(reader, "Taxa de rendimento (ex: 0.05): ")
			}

			dados := map[string]interface{}{
				"login": login, "nome": nome, "numero": numero, "tipo": tipo,
				"saldo_inicial": saldo, "limite": limite, "taxa": taxa,
			}
			res := enviarParaFila("criar_cliente", dados)
			fmt.Printf("\n=> [MQ] %v\n", res["mensagem"])

		case "4":
			numero := readInt(reader, "Número da conta: ")
			res := enviarParaFila("consultar_saldo", map[string]interface{}{"numero": numero})
			if res["sucesso"] == true {
				fmt.Printf("\n=> [MQ] Saldo atual: R$ %v (%v)\n", res["saldo"], res["tipo"])
			} else {
				fmt.Printf("\n=> [MQ] Erro: %v\n", res["mensagem"])
			}

		case "6":
			fmt.Println("Até logo!")
			return

		default:
			fmt.Println("=> Opção inválida (ou não implementada no atalho).")
		}
	}
}

File: cliente_js/cliente.js
import * as readline from 'node:readline/promises';
import { stdin as input, stdout as output } from 'node:process';
import amqplib from 'amqplib';
import crypto from 'crypto';

const rl = readline.createInterface({ input, output });

async function iniciarCliente() {
   
    const connection = await amqplib.connect('amqp://localhost');
    const channel = await connection.createChannel();

    
    const q = await channel.assertQueue('', { exclusive: true });

    
    const pendingRequests = new Map();

    
    channel.consume(q.queue, (msg) => {
        if (msg && pendingRequests.has(msg.properties.correlationId)) {
            const resolve = pendingRequests.get(msg.properties.correlationId);
            resolve(JSON.parse(msg.content.toString()));
            pendingRequests.delete(msg.properties.correlationId);
        }
    }, { noAck: true });

   
    async function enviarParaFila(acao, dados) {
        return new Promise((resolve) => {
            const correlationId = crypto.randomUUID();
            pendingRequests.set(correlationId, resolve);

            const payload = JSON.stringify({ acao, dados });

            channel.sendToQueue('fila_banco', Buffer.from(payload), {
                correlationId: correlationId,
                replyTo: q.queue 
            });
        });
    }

    async function menu() {
        console.log("\n--- BANCO RN (Cliente MQ / Node.js) ---");
        console.log("1 - Criar conta");
        console.log("2 - Depositar");
        console.log("3 - Sacar");
        console.log("4 - Consultar saldo");
        console.log("5 - Transferir");
        console.log("6 - Sair");

        const opcao = await rl.question("\nEscolha: ");

        if (opcao === "1") {
            const login = await rl.question("Login: ");
            const nome = await rl.question("Nome: ");
            const tipo = await rl.question("Tipo de conta (corrente/poupanca): ");
            const numero = await rl.question("Numero da conta: ");
            const saldo = await rl.question("Saldo inicial: ");
            
            let limite = 0, taxa = 0;
            if (tipo === "corrente") {
                limite = await rl.question("Limite cheque especial: ");
            } else {
                taxa = await rl.question("Taxa de rendimento (ex: 0.05): ");
            }

            const dados = {
                login, nome, numero: parseInt(numero), tipo,
                saldo_inicial: parseFloat(saldo), limite: parseFloat(limite), taxa: parseFloat(taxa)
            };

            const resposta = await enviarParaFila("criar_cliente", dados);
            console.log(`\n=> [MQ] ${resposta.mensagem}`);

        } else if (opcao === "2") {
            const numero = await rl.question("Número da conta: ");
            const valor = await rl.question("Valor para depositar: ");
            
            const resposta = await enviarParaFila("depositar", { numero: parseInt(numero), valor: parseFloat(valor) });
            console.log(`\n=> [MQ] ${resposta.mensagem}. ${resposta.sucesso ? 'Novo saldo: R$ ' + resposta.saldo : ''}`);

        } else if (opcao === "3") {
            const numero = await rl.question("Número da conta: ");
            const valor = await rl.question("Valor para sacar: ");
            
            const resposta = await enviarParaFila("sacar", { numero: parseInt(numero), valor: parseFloat(valor) });
            console.log(`\n=> [MQ] ${resposta.mensagem}. ${resposta.sucesso ? 'Novo saldo: R$ ' + resposta.saldo : ''}`);

        } else if (opcao === "4") {
            const numero = await rl.question("Número da conta: ");
            
            const resposta = await enviarParaFila("consultar_saldo", { numero: parseInt(numero) });
            if (resposta.sucesso) {
                console.log(`\n=> [MQ] Saldo atual: R$ ${resposta.saldo} (${resposta.tipo})`);
            } else {
                console.log(`\n=> [MQ] Erro: ${resposta.mensagem}`);
            }

        } else if (opcao === "5") {
            const origem = await rl.question("Sua conta (Origem): ");
            const destino = await rl.question("Conta destino: ");
            const valor = await rl.question("Valor para transferir: ");
            
            const resposta = await enviarParaFila("transferir", {
                origem: parseInt(origem), destino: parseInt(destino), valor: parseFloat(valor)
            });
            console.log(`\n=> [MQ] ${resposta.mensagem}. ${resposta.sucesso ? 'Seu saldo: R$ ' + resposta.saldo_origem : ''}`);

        } else if (opcao === "6") {
            console.log("Até logo");
            await channel.close();
            await connection.close();
            rl.close();
            process.exit(0);
        } else {
            console.log("\n=> Opção inválida.");
        }

        if (opcao !== "6") await menu();
    }

    menu();
}

iniciarCliente().catch(console.error);
